rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java

```diff
@@ -506,172 +510,150 @@ public void onFailure(Throwable caught) {
+     * It it is regular, than rm portal should treat server response as a delta.
+     * If it is not regular, than rm portal should 'forget' what it knew, and start from the scratch
+     * @param currentCounter is current counter that client is aware of
+     * @param latestCounter is counter received from server
+     * @return true if request is regular
+     */
+    private boolean isRegularRequest(long currentCounter, Long latestCounter) {
+        return 0 <= currentCounter && currentCounter <= latestCounter;
+    }
+
+    /**
+     * Add and remove NodeSources
+     * @param newNodeSources
+     * @param obj
+     */
+    private void processNodeSources(HashMap<String, NodeSource> newNodeSources, JSONObject obj) {
+        JSONArray jsNodeSources = obj.get("nodeSource").isArray();
+        for (int i = 0; i < jsNodeSources.size(); i++) {
+            JSONObject jsNodeSource = jsNodeSources.get(i).isObject();
+
+            NodeSource nodeSource = parseNodeSource(jsNodeSource);
+            if (nodeSource.isRemoved()) {
+                newNodeSources.remove(nodeSource.getSourceName());
+            } else {
+                newNodeSources.put(nodeSource.getSourceName(), nodeSource);
+            }
+        }
+    }
+
+    private void removeNodeFromNodeSource(Node node, NodeSource nodeSource) {
+        if (node.isDeployingNode()) {
+            nodeSource.getDeploying().remove(node.getNodeUrl());
+        } else {
+            Host host = nodeSource.getHosts().get(node.getHostName());
+
+            if (host != null) {
+                host.getNodes().remove(node.getNodeUrl());
+
+                if (host.getNodes().isEmpty()) {
+                    nodeSource.getHosts().remove(host);
+                }
+            }
+        }
+    }
 
+    private void addNodeToNodeSource(Node node, NodeSource nodeSource) {
+        // as deploying node
+        if (node.isDeployingNode()) {
+
+            nodeSource.getDeploying().put(node.getNodeUrl(), node);
+
+        } else { // as already deployed node
+            Host host = nodeSource.getHosts().get(node.getHostName());
+
+            if (host == null) { // create host if there is no host
+                host = new Host(node.getHostName(), node.getSourceName());
+                nodeSource.getHosts().put(node.getHostName(), host);
+            }
+
+            host.getNodes().put(node.getNodeUrl(), node);
+
+            if (node.isVirtual()) {
+                host.setVirtual(true);
+            }
+        }
+    }
+
+    /**
+     * @return clones node sources with all hosts and nodes
+     */
+    private void copyNodesSources(Map<String, NodeSource> oldNodeSources, HashMap<String, NodeSource> newNodeSources) {
+        for (NodeSource nodeSource : oldNodeSources.values()) {
+            NodeSource newNodeSource = new NodeSource(nodeSource);
+            newNodeSources.put(newNodeSource.getSourceName(), newNodeSource);
+        }
+    }
+
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1181)